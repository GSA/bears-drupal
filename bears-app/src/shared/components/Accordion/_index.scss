.usa-accordion {
  &__button {
    position: relative;

    img {
      position: absolute;
      right: 30px;
      top: 30px;
    }

    &[aria-expanded='true'] {
      background-image: none;
      background-color: #1a4480;
      color: #fff;
    }

    &[aria-expanded='false'] {
      background-image: none;
      background-color: #d9e8f6;
      color: #1a4480;
    }

    .title {
      font-size: 20px;
      font-weight: 700;
      line-height: 27px;
      letter-spacing: 0;
    }

    .title-description {
      font-size: 16px;
      font-weight: 400;
      line-height: 22px;
      letter-spacing: 0;
    }
  }

  &__content {
    border-bottom: 1px solid #005ea2;
    border-left: 1px solid #005ea2;
    border-right: 1px solid #005ea2;
    padding-bottom: 1rem;
  }
}
