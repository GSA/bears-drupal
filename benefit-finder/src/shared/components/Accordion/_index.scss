@use '../../styles/colors/_index.scss' as color;
@use '../../styles/mixins/_index.scss' as *;
@use '../../styles/space/_index.scss' as space;

#benefit-finder {
  .benefit-accordion {
    .usa-accordion__button {
      position: relative;

      svg {
        position: absolute;
        top: 0;
        bottom: 0;
        right: 20px;
        margin: auto 0;
      }

      &[aria-expanded='true'] {
        background-image: none;
        background-color: color.$marine;
        color: color.$white;
      }

      &[aria-expanded='false'] {
        background-image: none;
        background-color: color.$sky;
        color: color.$marine;
      }

      .heading {
        @include h4;

        letter-spacing: 0;
      }

      .sub-heading {
        @include h5;

        letter-spacing: 0;
      }
    }

    .usa-accordion__content {
      border-bottom: 1px solid color.$pop-blue;
      border-left: 1px solid color.$pop-blue;
      border-right: 1px solid color.$pop-blue;
      padding-bottom: space.$space-md;
    }
  }
}
