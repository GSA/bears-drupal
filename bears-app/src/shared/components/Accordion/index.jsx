import { useEffect, useState } from 'react'
import Close from './assets/close.svg'
import Open from './assets/open.svg'
import PropTypes from 'prop-types'
import './_index.scss'

/**
 * a functional component that renders a usa accordion
 * @component
 * @param {string} id - inherited id
 * @param {string} heading - title value
 * @param {string} subHeading - eligibleStatus value
 * @param {React.ReactNode} children - inherited children
 * @param {bool} isExpanded - sets the open state of an accordion
 * @return {html} returns a semantic html button element
 */
const Accordion = ({
  id,
  heading,
  subHeading,
  children,
  isExpanded,
  ...props
}) => {
  /**
   * a hook that hanldes our open state of the accordion
   * @function
   * @return {boolean} returns true or false
   */
  const [isOpen, setOpen] = useState(false)

  useEffect(() => {
    setOpen(isExpanded)
  }, [isExpanded])

  /**
   * a handler that returns the proper icon
   * @function
   * @param {React.ReactNode} children - inherited children
   * @return {html} returns a semantic img element with proper svg
   */
  const handleIcon = () =>
    !isOpen ? (
      <img src={Open} alt="a plus icon" />
    ) : (
      <img src={Close} alt="a minus icon" />
    )

  return (
    <div className="benefit-accordion" {...props}>
      {/* we don't use `usa-accordion` class because it is too opinionated about control, this throws an error from the uswds javascript, but does not impact/break functionality */}
      <h4 className="usa-accordion__heading">
        <button
          type="button"
          className="usa-accordion__button"
          aria-expanded={isOpen || false}
          aria-controls={id}
          onClick={() => setOpen(!isOpen)}
        >
          <span className="heading">{heading}</span>
          <br />
          <span className="sub-heading">{subHeading}</span>
          {handleIcon()}
        </button>
      </h4>
      <div
        id={id}
        className="usa-accordion__content usa-prose"
        hidden={!isOpen}
      >
        <div>{children}</div>
      </div>
    </div>
  )
}

Accordion.propTypes = {
  id: PropTypes.string,
  heading: PropTypes.string,
  subHeading: PropTypes.string,
  description: PropTypes.string,
  children: PropTypes.node,
}

export default Accordion
