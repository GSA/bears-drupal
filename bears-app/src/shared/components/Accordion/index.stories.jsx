import Accordion from './index.jsx'

export default {
  component: Accordion,
  tags: ['autodocs'],
  args: {
    id: 1,
    title: 'First Amendment',
    description: 'Likely Eligible',
    children:
      'Congress shall make no law respecting an establishment of religion, or prohibiting the free exercise thereof; or abridging the freedom of speech, or of the press; or the right of the people peaceably to assemble, and to petition the Government for a redress of grievances.',
  },
}

export const Primary = {}
