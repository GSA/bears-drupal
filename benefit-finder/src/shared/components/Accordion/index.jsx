import { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import './_index.scss'

/**
 * a functional component that renders a usa accordion
 * @component
 * @param {string} id - inherited id
 * @param {string} heading - title value
 * @param {string} subHeading - eligibleStatus value
 * @param {React.ReactNode} children - inherited children
 * @param {bool} isExpanded - sets the open state of an accordion
 * @return {html} returns a semantic html button element
 */
const Accordion = ({
  id,
  heading,
  subHeading,
  children,
  isExpanded,
  hidden,
  ...props
}) => {
  /**
   * a hook that hanldes our open state of the accordion
   * @function
   * @return {boolean} returns true or false
   */
  const [isOpen, setOpen] = useState(false)

  useEffect(() => {
    setOpen(isExpanded)
  }, [isExpanded])

  const Close = () => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path d="M0 0h24v24H0z" fill="none" />
      <path d="M19 13H5v-2h14v2z" fill="#ffffff" />
    </svg>
  )

  const Open = () => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path d="M0 0h24v24H0z" fill="none" />
      <path fill="#1a4480" d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />
    </svg>
  )

  /**
   * a handler that returns the proper icon
   * @function
   * @param {React.ReactNode} children - inherited children
   * @return {html} returns a semantic img element with proper svg
   */
  const handleIcon = () =>
    !isOpen ? <Open alt="a plus icon" /> : <Close alt="a minus icon" />

  return (
    <div className="benefit-accordion" {...props} hidden={hidden}>
      {/* we don't use `usa-accordion` class because it is too opinionated about control, this throws an error from the uswds javascript, but does not impact/break functionality */}
      <h4 className="usa-accordion__heading">
        <button
          type="button"
          className="usa-accordion__button"
          aria-expanded={isOpen || false}
          aria-controls={id}
          onClick={() => setOpen(!isOpen)}
        >
          <span className="heading">{heading}</span>
          <br />
          <span className="sub-heading">{subHeading}</span>
          {handleIcon()}
        </button>
      </h4>
      <div
        id={id}
        className="usa-accordion__content usa-prose"
        hidden={!isOpen}
      >
        <div>{children}</div>
      </div>
    </div>
  )
}

Accordion.propTypes = {
  id: PropTypes.string,
  heading: PropTypes.string,
  subHeading: PropTypes.string,
  description: PropTypes.string,
  children: PropTypes.node,
}

export default Accordion
