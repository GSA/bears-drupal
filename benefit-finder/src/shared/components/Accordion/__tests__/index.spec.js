import renderer from 'react-test-renderer'
import Accordion from '../index.jsx'

describe('Accordion', () => {
  it('renders a match to the previous snapshot', () => {
    const component = renderer.create(<Accordion />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
