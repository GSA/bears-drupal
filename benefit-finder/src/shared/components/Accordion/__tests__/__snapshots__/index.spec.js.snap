// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders a match to the previous snapshot 1`] = `
<div
  className="benefit-accordion usa-accordion"
>
  <h4
    className="usa-accordion__heading"
  >
    <button
      aria-expanded={false}
      className="usa-accordion__button"
      onClick={[Function]}
      type="button"
    >
      <span
        className="heading"
      />
      <br />
      <span
        className="sub-heading"
      />
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          fill="#1a4480"
        />
      </svg>
    </button>
  </h4>
  <div
    className="usa-accordion__content usa-prose"
    hidden={true}
  >
    <div />
  </div>
</div>
`;
